hashing.go

```diff
@@ -0,0 +1,170 @@
+//go:build go1.18
+// +build go1.18
+
+package xsync
+
+import (
+	"hash/maphash"
+	"reflect"
+	"unsafe"
+)
+
+// makeHashFunc creates a fast hash function for the given comparable type.
+// The only limitation is that the type should not contain interfaces inside
+func makeHashFunc[T comparable]() func(maphash.Seed, T) uint64 {
+	var zero T
+	rt := reflect.TypeOf(&zero).Elem() // Elem() avoids uninformative panics when T is interface
+
+	blocks := hGetMemLayout(rt)
+
+	// empty struct case, should never happen irl
+	if len(blocks) == 0 {
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.Bytes(seed, nil)
+		}
+	}
+
+	// string or a struct with a single string field
+	if len(blocks) == 1 && !blocks[0].IsString {
+		block := blocks[0]
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.Bytes(seed, asBytes(unsafe.Pointer(&v), block.Offset, block.Length))
+		}
+	}
+
+	// data type is a single contiguous region of memory (e.g. struct with integer fields)
+	if len(blocks) == 1 && blocks[0].IsString {
+		block := blocks[0]
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.String(seed, asString(unsafe.Pointer(&v), block.Offset))
+		}
+	}
+
+	// complex data type consisting of multiple blocks
+	{
+		return func(seed maphash.Seed, v T) uint64 {
```

Comment:
Looks like this implementation is prone to hash collision attacks. E.g.:
```
type Person struct {
    FirstName, LastName string
}
```

With such struct instances like `Person{"Foo", "Bar"}` and `Person{"Fo", "oBar"}` will have the same hash code value. This example tries to avoid that in polynomial hash function style: https://github.com/puzpuzpuz/xsync/blob/c3b502018e15a1ef54a4e2ced12a1c392746a5bc/example_test.go#L21-L32

I didn't check the standard hash function, but most likely it prevents such attacks. WDYT?

Reply:
Built in hash function does it field-by-filed and result of the previous hashing is chained into the next one.
Actually, there is the typehash function in https://go.dev/src/runtime/alg.go that illustrates this behavior.

This, by the way, is not consistent with stdlib's maphash that does hashing in 128byte chunks.

I, on the other hand, tried to use as large chunks as possible, but as you said that's not resistant to hash collision attacks.

So, yes, you're right. My code is bad in terms of hash collision attacks. I tried to concatenate as much fields as possible and to minimize calls of hash function. I though I was optimizing the code, but I was wrong.